### ISSUE TITLE
Error when creating worktrees that contain slashes

### ISSUE THREAD

**Issue by hakoerber** [libgit2/libgit2#6327]
I experienced surprising errors when using git worktrees that contain slashes.

### Reproduction steps

I wrote a script outlining the behaviour here: https://github.com/hakoerber/libgit2-worktree-mre/blob/master/run.sh
The respective libgit2 code: https://github.com/hakoerber/libgit2-worktree-mre/blob/master/main.c

TL;DR:

Calling `git_worktree_add()` with a worktree containing slashes fails, as libgit2 does not create necessary directories under `.git/worktrees` and the base directories for the worktree.

Example:

```c
git_worktree_add(&worktree, repo, "a/foo", "a/foo", NULL);
```
This will lead to the following error:

`failed to make directory '/tmp/example/.git/worktrees/a/foo': No such file or directory`

When creating `.git/worktrees/a` manually, it will still fails, this time because libgit2 cannot create the worktree directory:

`failed to make directory 'a/foo': No such file or directory`.

When creating `a` manually, it works.

### Anaylsis & possible fix

I traced this down to the usage of `git_futils_mkdir()` [here](https://github.com/libgit2/libgit2/blob/f98dd5438f8d7bfd557b612fdf1605b1c3fb8eaf/src/libgit2/worktree.c#L348) (for `.git/worktrees`) and [here](https://github.com/libgit2/libgit2/blob/f98dd5438f8d7bfd557b612fdf1605b1c3fb8eaf/src/libgit2/worktree.c#L354) (for the workdir). AFAICT, `git_futils_mkdir()` does not create parent directories (unless told to), so the call will fail.

When adding `GIT_MKDIR_PATH` to the options of `git_futils_mkdir` here, everything works as expected.

Note that the git CLI allows worktrees with slashes (even though the handling of `.git/worktrees` is surprising, as outlined in the reproduction script).

---

Before I invest time in a fix, I'd like to check with you whether this is something that *should* be fixed (i.e. whether you want to support worktrees with slashes), and how to best fix it.

I don't think just adding `GIT_MKDIR_PATH` is a good solution, as it may lead to potential problems (e.g. when using `foo/../../../../../../../bar` as the worktree name). We'd have to make sure to only create strict subdirectories of `.git/worktrees` or the current directory. My C is too rusty to find a good solution with `git_futils_mkdir()`, so I'd be happy if you could advise on the best appraoch in case you want to go this route.

### Expected behavior

When using worktrees that contain slashes, I expect libgit2 to create a worktree directory successfully.

### Actual behavior

libgit2 fails with the following error, when using `foo/bar` as the worktree name:

```
failed to make directory '/example/.git/worktrees/foo/bar': No such file or directory
```

### Version of libgit2 (release number or SHA1)

`f98dd5438f8d7bfd557b612fdf1605b1c3fb8eaf` (i.e. current (2022-06-16T22:27:31) `main`).

### Operating system(s) tested

Linux / Ubuntu 20.04.

**Comment by ethomson** [comment]
Thanks for the report!

>  I'd like to check with you whether this is something that should be fixed (i.e. whether you want to support worktrees with slashes),

Is there a way to create a worktree _name_ with a slash in it using git? For example, if I do:

```
git worktree add ../z/z/z/z/z
```

Then my worktree is created in `../z/z/z/z/z` but the data is created at `.git/worktrees/z`. (Not `z/z/z/z/z`...)

Can git support worktrees nested beneath other directories? It looks like *no*. If I manually update a repository with this structure, it fails.

My naive assumption is to reject worktree names with `/`, but I'm relatively new to _using_ worktrees, let alone their low-level implementation, so I may be misunderstanding.

**Comment by hakoerber** [comment]
Upon further reading, I think there is no such thing as a worktree *name*. There are two things that are relevant for a worktree:

* The path that it is checked out in. This is the parameter to `git worktree add`.
* The branch that is checked out in the worktree. According to `git-worktree(1)`, this defaults to `$(basename <path>)`. But you can override this with `-b`:

```bash
$ cd $(mktemp -d) && git init >/dev/null && git commit --allow-empty -m "Initial commit" >/dev/null
$ git worktree add -b z/z/z/z z/z/z/z
Preparing worktree (new branch 'z/z/z/z')
HEAD is now at 78940fc9 Initial commit
$ git worktree list
/tmp/tmp.123          78940fc9 [master]
/tmp/tmp.123/z/z/z/z  78940fc9 [z/z/z/z]
$ 
```

Note that the directory in `.git/worktrees` is just `z`:

```
$ ls -1 .git/worktrees/z
HEAD
ORIG_HEAD
commondir
gitdir
index
logs
$
```

This matches [the docs](https://git-scm.com/docs/git-worktree#_details):

> The private sub-directory’s name is usually the base name of the linked worktree’s path, possibly appended with a number to make it unique.

So if you create a different worktree that has the same basename:

```
$ git worktree add -b a/z/z/z a/z/z/z
Preparing worktree (new branch 'a/z/z/z')
HEAD is now at 77f5eba7 Initial commit
$ ls -1 .git/worktrees
z
z1
$
```

As far as I can see, **the name of the directory in `.git/worktrees` just does not matter**. Git just goes through all first-level directories in `.git/worktrees`. It can determine the path of the worktree from `gitdir` and check which branch is checked out.

I tested this:

```
$ mv .git/worktrees/z .git/worktrees/foobar
$ git worktree list 
/tmp/tmp.123          619a0560 [master]
/tmp/tmp.123/a/z/z/z  619a0560 [a/z/z/z]
/tmp/tmp.123/z/z/z/z  619a0560 [z/z/z/z]
```

As you can see, the worktrees still show up as before.

In summary, having worktrees within subdirectories is valid and should actually be transparent to the worktree handling code. The only thing missing to `libgit2` is the handling of conflicts inside `.git/worktrees`).

My initial suggestion (just creating nested subdirectories under `.git/worktrees`) is incorrect. `git` will not find those kind of worktrees. The relevant code can be found [here](https://github.com/git/git/blob/30cc8d0f147546d4dd77bf497f4dec51e7265bd8/worktree.c#L132). Note that `readdir_skip_dot_and_dotdot()` does not recurse.

So if we decide to support worktrees inside subdirectories, I propose that we implement the documented behaviour:

* When creating a worktree with slashes, try to create `.git/worktrees/$(basebae $name)`. If that one does not exist:
  * Try to create `.git/worktrees/$(basebae $name)$i`, with `i = 1`.
  * Each time this fails, increment `i` and try again.

---

What do you think? Do you want to have this behaviour in `libgit2`?

If yes, how do you want to proceed? Should I just give it a shot and open a PR?

**Comment by ethomson** [comment]
> As far as I can see, the name of the directory in .git/worktrees just does not matter.

I agree. I don't know if this was a git implementation detail that's changed or if it was never really relevant. I would suggest that we accept `NULL` for name, document that we suggest people _use_ `NULL`, and then use your proposal for deriving the name when a `NULL` is provided:

> * When creating a worktree with slashes, try to create .git/worktrees/$(basebae $name). If that one does not exist:
>     * Try to create .git/worktrees/$(basebae $name)$i, with i = 1.
>     * Each time this fails, increment i and try again.

If a non-`NULL` is provided for name, we should not try to mutate it. We should sanity check it for legal characters (ie, reject `/`) and not do any conflict resolution, if there's an existing one we should fail.

The providing-a-name option then becomes a power user feature that we recommend people avoid.

How does that sound?

**Comment by ethomson** [comment]
> That also means that the change is backwards compatible, as NULL is currently not a valid value for name. Also, any name containing / will currently fail (which is why this issue exists in the first place), so it's fine to reject this explictly.

🎉  Excellent.

> If it's non-NULL, we will retain the current behaviour.

I think that we should explicitly reject bogus names if we don't already. It's possible that we do, I'm not certain offhand. But I can imagine us tightening up our restrictions.

> If the directory contains slashes, we try to create .git/worktrees/${name}, falling back to appending numbers if the directory already exists --- as described above.

Do you mean `${basename(name)}` here? That might have been implicit, apologies if so.

**Comment by hakoerber** [comment]
> I think that we should explicitly reject bogus names if we don't already. It's possible that we do, I'm not certain offhand. But I can imagine us tightening up our restrictions.

I just checked spaces for example: Having a worktree in a subdirectory with spaces is fine. Only the branch cannot contain spaces. So something like `git worktree add "with spaces"` breaks, but `git worktree add -b foo "with spaces"` works. I *guess* everything that is a valid path is ok (so everything except the null byte on Linux, while Windows will most likely choke on something like `:`).

I'll check the current restrictions. But we have to keep in mind that this means that we break backwards compatibility. Spaces are for example also possible in libgit2: `git_worktree_add(&worktree, repo, "foo", "with spaces", NULL)` works.

I guess it's up to you do decide on backwards compatibilty.

> Do you mean ${basename(name)} here? That might have been implicit, apologies if so.

Yes, sorry, oversight on my part. I edited it.
